src/main/java/picard/illumina/parser/NewIlluminaDataProvider.java

```diff
@@ -0,0 +1,116 @@
+     */
+    NewIlluminaDataProvider(final List<File> cbcls, final List<AbstractIlluminaPositionFileReader.PositionInfo> locs,
+                            final File[] filterFiles, final int lane, final int tileNum, final OutputMapping outputMapping,
+                            boolean usingQualityScores, final Map<String, BarcodeMetric> barcodesMetrics,
+                            final BarcodeMetric noMatchMetric,
+                            int maxNoCalls, int maxMismatches, int minMismatchDelta, int minimumBaseQuality) {
+        super(lane, outputMapping);
+        this.usingQualityScores = usingQualityScores;
+        this.barcodeMetricMap = barcodesMetrics;
+        this.maxNoCalls = maxNoCalls;
+        this.maxMismatches = maxMismatches;
+        this.minMismatchDelta = minMismatchDelta;
+        this.minimumBaseQuality = minimumBaseQuality;
+        Map<Integer, File> filterFileMap = new HashMap<>();
+        for (File filterFile : filterFiles) {
+            filterFileMap.put(fileToTile(filterFile.getName()), filterFile);
+        }
+        this.reader = new CbclReader(cbcls, filterFileMap, outputMapping.getOutputReadLengths(), tileNum, locs);
+        this.outputReadStructure = outputMapping.getOutputReadStructure();
+
+        this.noMatchMetric = noMatchMetric;
+    }
+
+    @Override
+    public void close() {
+        reader.close();
+    }
+
+    @Override
+    public void seekToTile(int seekAfterFirstRead) {
+
+    }
+
+    @Override
+    public boolean hasNext() {
+        return reader.hasNext();
+    }
+
+    @Override
+    public ClusterData next() {
+        CbclData cbclData = reader.next();
+
+        if (cbclData == null) return null;
+
+        final ClusterData cluster = new ClusterData(outputReadTypes);
+        cluster.setLane(lane);
+        cluster.setTile(cbclData.getTile());
+
+        final int[] barcodeIndices = outputReadStructure.sampleBarcodes.getIndices();
+        addReadData(cluster, numReads, cbclData);
+
+        final byte[][] barcodeSubsequences = new byte[barcodeIndices.length][];
+        final byte[][] qualityScores = usingQualityScores ? new byte[barcodeIndices.length][] : null;
+        for (int i = 0; i < barcodeIndices.length; i++) {
+            barcodeSubsequences[i] = cluster.getRead(barcodeIndices[i]).getBases();
+            if (usingQualityScores) qualityScores[i] = cluster.getRead(barcodeIndices[i]).getQualities();
+        }
+
+
+        final BarcodeMatch match = BarcodeMatch.findBestBarcodeAndUpdateMetrics(barcodeSubsequences, qualityScores,
+                true, barcodeMetricMap, noMatchMetric, maxNoCalls, maxMismatches,
+                minMismatchDelta, minimumBaseQuality);
+
+        if (match.isMatched()) {
+            cluster.setMatchedBarcode(match.getBarcode());
+        }
+        return cluster;
+    }
+
+    private Integer fileToTile(final String fileName) {
+        final Matcher matcher = Pattern.compile("^s_\\d+_(\\d{1,5}).+").matcher(fileName);
```

Comment:
Another `Pattern.compile()` that could be done statically. Putting it in a static also provides an opportunity to document it, since most regular expressions are harder to read than they are to write.